src/backend/distributed/progress/multi_progress.c

```diff
@@ -0,0 +1,287 @@
+/*-------------------------------------------------------------------------
+ *
+ * multi_progress.c
+ *	  Routines for tracking long-running jobs and seeing their progress.
+ *
+ * Copyright (c) 2017, Citus Data, Inc.
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "miscadmin.h"
+#include "pgstat.h"
+
+#include "distributed/multi_logical_optimizer.h"
+#include "distributed/multi_progress.h"
+#include "nodes/execnodes.h"
+#include "storage/dsm.h"
+#include "utils/builtins.h"
+
+
+/* Dynamic shared memory handle of the current progress */
+static uint64 currentProgressDSMHandle = DSM_HANDLE_INVALID;
+
+
+static ReturnSetInfo * FunctionCallGetTupleStore1(PGFunction function, Oid functionId,
+												  Datum argument);
+
+
+/*
+ * CreateProgressMonitor is used to create a place to store progress information related
+ * to long running processes. The function creates a dynamic memory segment consisting of
+ * a header regarding to the process and an array of "steps" that the long running
+ * "operations" consists of. The handle of the dynamic shared memory is stored in
+ * pg_stat_get_progress_info output, to be parsed by a progress retrieval command
+ * later on. This behavior may cause unrelated (but hopefully harmless) rows in
+ * pg_stat_progress_vacuum output. The caller of this function should provide a magic
+ * number, a unique 64 bit unsigned integer, to distinguish different types of commands.
+ */
+void *
+CreateProgressMonitor(uint64 progressTypeMagicNumber, int stepCount, Size stepSize,
+					  Oid relationId)
+{
+	dsm_segment *dsmSegment = NULL;
+	dsm_handle dsmHandle = 0;
+	void *dsmSegmentAddress = NULL;
+	ProgressMonitorHeader *header = NULL;
+	void *steps = NULL;
+	Size monitorSize = 0;
+
+	if (stepSize <= 0 || stepCount <= 0)
+	{
+		ereport(ERROR,
+				(errmsg("Number of steps and size of each step should be positive "
+						"values.")));
+	}
+
+	monitorSize = sizeof(ProgressMonitorHeader) + stepSize * stepCount;
+	dsmSegment = dsm_create(monitorSize, DSM_CREATE_NULL_IF_MAXSEGMENTS);
+
+	if (dsmSegment == NULL)
+	{
+		ereport(WARNING,
+				(errmsg(
+					 "Couldn't create a dynamic shared memory segment to keep track of "
+					 "progress of the current command.")));
```

Comment:
`errmsg` convention is to use `could not create ...` and no dot at the end.

Maybe shorten the lines here such that `errmsg` fits on the same line as `ereport`.